lib/modules/manager/cargo/artifacts.ts

```diff
@@ -29,6 +29,27 @@ async function cargoUpdate(
   await exec(cmd, execOptions);
 }
 
+async function cargoUpdatePackage(
+  manifestPath: string,
+  depName: string,
+  lockedVersion: string,
+  newVersion: string | undefined,
+  constraint: string | undefined
+): Promise<void> {
+  let cmd = `cargo update --manifest-path ${quote(manifestPath)}`;
+  cmd += ` --package ${depName}@${lockedVersion}`;
+
+  if (newVersion) {
+    cmd += ` --precise ${newVersion}`;
+  }
```

Comment:
I could see this fail:
-  we depend on `foo` with  `2.0.1`
- the latest version from crates.io is detected as `2.0.2`
- another package in `Cargo.lock` depends on `foo` with  `=2.0.1` (exact version)

Renovate will try and update `foo` to `2.0.2` at which point `cargo update` will fail saying there's no way to resolve the dependencies given these constraints.

For this to work, the cargo versioning module would need to be aware of `Cargo.lock` as well.

Reply:
As predicted, I started testing this on some real-world repos and ran into this exact issue:
```
Updating crates.io index
error: failed to select a version for `wasm-bindgen`.
    ... required by package `js-sys v0.3.64`
    ... which satisfies dependency `js-sys = "^0.3"` of package `nickel-lang-core v0.1.0`
    ... which satisfies path dependency `nickel-lang-core` (locked to 0.1.0) of package `nickel-lang-cli v1.0.0`
versions that meet the requirements `^0.2.87` are: 0.2.87
all possible versions conflict with previously selected packages.
  previously selected package `wasm-bindgen v0.2.83`
    ... which satisfies dependency `wasm-bindgen = "=0.2.83"` of package `nickel-lang-core v0.1.0`
    ... which satisfies path dependency `nickel-lang-core` (locked to 0.1.0) of package `nickel-lang-cli v1.0.0`
failed to select a version for `wasm-bindgen` which could resolve this conflict
```

Any ideas what we should do in cases where we can't upgrade to `newVersion`? Do we need to adjust the versioning module to take the lock file into account? Is that even possible with the current API?

Reply:
So first of all, it's perfectly OK to raise a PR but with failing artifacts. Renovate makes it very clear using both a failing check run as well as a PR comment. Users have the ability to suppress failing PRs if they really wish, but usually it's better to know.

However it's also OK for you to add an algorithm to catch and retry/refine the update if the output from cargo is clear enough. e.g. try updating only the selected dependency/dependencies first, but then update more of them in another retry. 

I did something similar for bundler here: https://github.com/renovatebot/renovate/blob/76c9c63417d72e38dac7d00040c33186f9df4dad/lib/modules/manager/bundler/artifacts.ts#L248-L272

Reply:
I think for the initial version I'd rather just open a PR with failing artifacts. The issue is that this instance this will most likely result in an empty commit, so no PR will actually be created. I added additional logging to `cargoUpdatePrecise` so it's more obvious which package failed to update and why.